### PR TITLE
Migrate keyboards using uGFX to LED_MATRIX

### PR DESCRIPTION
**Pull request by firetech**
Migrate Input Club keyboards (Ergodox Infinity and Whitefox) away from uGFX for keyboard backlight.

## Description

Ergodox Infinity and Whitefox use uGFX to control their keyboard backlight. Because of licensing issues, it's desirable to move away from uGFX (see #5268). The chip in question (IS31FL3731C) already had a working driver for use with LED Matrix (drivers/issi/is31fl3731-simple.c).

Since uGFX is still compiled for the LCD display in Ergodox Infinity,  and it caused some naming conflicts, I had to rename the `point_t` type of LED/RGB Matrix. `led_point_t` seemed fitting. (This is in conflict with two lines in #12651, but is an easy fix.)

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by drashna** [comment]
I suspect that it will need more changes to get this working properly (read: 100%), but it looks like a good start, at least!

**Comment by tzarc** [comment]
Given that it's changing a bunch of things at once, it'll need to go to `develop` instead of `master`.

**Comment by firetech** [comment]
> I suspect that it will need more changes to get this working properly (read: 100%), but it looks like a good start, at least!

Definitely. For now, I'm waiting for the i2c support to be merged before doing more work on this.

**Comment by drashna** [comment]
Looks like there is an issue with the whitefox board: 
```
Making whitefox with keymap default                                                                    [ERRORS]
lib/ugfx/src/gdisp/gdisp.c: In function '_gdispInit':
lib/ugfx/src/gdisp/gdisp.c:566:4: error: useless type name in empty declaration [-Werror]
    extern GDISPVMTLIST        GDISP_DRIVER_LIST;
    ^~~~~~
cc1: all warnings being treated as errors
 | 
 | 
 | 
make[1]: *** [tmk_core/rules.mk:386: .build/obj_whitefox/lib/ugfx/src/gdisp/gdisp.o] Error 1
make[1]: *** Waiting for unfinished jobs....
```

**Comment by firetech** [comment]
> Looks like there is an issue with the whitefox board:

Yeah, I haven't touched the Whitefox yet, but I have changed some common code in quantum visualizer, so it's entirely expected that it doesn't compile yet. ;)

Making this posting as a PR might have been jumping the gun a bit, it might have been better to make a normal issue and just refer to my branch...

**Comment by fauxpark** [comment]
The I2C PR has been merged now, so this should be good to progress.
Additionally, I just discovered the Infinity60 (also by Input:Club) uses the 3731C too, but has some custom controller code instead of uGFX.

**Comment by skullydazed** [comment]
@firetech do you think you'll be working on this again?

**Comment by tzarc** [comment]
I... have no idea why I didn't review this previously.
I've sorted out conflicts as best I could -- would be great to get things validated.
Would prefer to get the `#TODO: Remove` stuff removed for the PR merge, though.

**Comment by fauxpark** [comment]
I'm happy to report that I have LED matrix more or less caught up with RGB matrix and ported over a couple animations to test. I just need to break down the giant diff into more digestible PRs.

Should make this PR much more useful :)

**Comment by firetech** [comment]
> @firetech do you think you'll be working on this again?

Maybe, but probably not in the near future. Life has been lacking free time and motivation lately... If someone else would like to continue the work, be my guest.

**Comment by firetech** [comment]
This is almost working properly now. The LED positions in `g_led_config` are all zero (I'll use the data in info.json to calculate proper positions at some point), and I'm having some issues with `led_matrix_eeconfig.flags` misbehaving (see discussion in #12580).

However, my keyboard is behaving slightly random when it comes to waking up after suspend (sometimes not waking at all, sometimes waking but not calling `suspend_wakeup_init()`, and sometimes working properly), but I think it's the same on master currently (I haven't updated my daily driver firmware in a while)...

**Comment by firetech** [comment]
Forgot to mention that the newest code no longer breaks the visualizer since LED Matrix has been decoupled from the Backlight system. Both old and new models of LED control are therefore (hopefully) supported (but will probably break if you try to run both at the same time, of course).

This also should make it possible to merge this without porting Whitefox.

**Comment by firetech** [comment]
I think this is (finally!) ready for review. 😄

**Comment by fauxpark** [comment]
#12845 may help with that.

**Review by drashna — APPROVED** [review]

**Comment by firetech** [comment]
Since #12651 was merged, I'll have to do a small update of this to make it compileable again. Don't merge this yet!

**Comment by fauxpark** [comment]
@firetech I've opened the above linked PR just to cut down on this PR's scope. The SDB stuff I want to handle differently, by simply specifying a pin in config.h and the driver will do the needful (including possibly deasserting it when the user toggles the LED/RGB matrix off etc to save power).

**Comment by firetech** [comment]
> @firetech I've opened the above linked PR just to cut down on this PR's scope. The SDB stuff I want to handle differently, by simply specifying a pin in config.h and the driver will do the needful (including possibly deasserting it when the user toggles the LED/RGB matrix off etc to save power).

Sounds reasonable. I mostly made this quick fix since the driver didn't handle it at all previously (when I started writing the patch last summer, no board using LED Matrix and IS31FL3731 seemed to have SDB connected to an MCU pin).

**Comment by fauxpark** [comment]
> @fauxpark Do you intend to work on that, or what is the plan?

Yup, but it might be a short while as technically develop is now locked to new PRs (apart from urgent bugfixes, of course).

**Comment by firetech** [comment]
Updated to be based on #12864 (not including it, but can't be built without it), and moved the IS31FL3731 SDB stuff to `keyboard_pre_init_kb()` (`matrix_init_kb()` is called after `led_matrix_init()`, so that didn't work), in anticipation of an improved, common way of setting SDB.

File: keyboards/whitefox/whitefox.c
Comment (fauxpark) [inline keyboards/whitefox/whitefox.c:23]:
```suggestion
```
These includes aren't needed anymore.

File: keyboards/whitefox/whitefox.c
Comment (fauxpark) [inline keyboards/whitefox/whitefox.c:75]:
```suggestion
    setPinOutput(B16);
    writePinHigh(B16);
```
Does this not work?

File: keyboards/ergodox_infinity/ergodox_infinity.c
Comment (fauxpark) [inline keyboards/ergodox_infinity/ergodox_infinity.c:301]:
```suggestion
```

File: keyboards/ergodox_infinity/ergodox_infinity.c
Comment (fauxpark) [inline keyboards/ergodox_infinity/ergodox_infinity.c:15]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by firetech** [comment]
Fixed the changes requested by @fauxpark, and removed some code made unnecessary by #12878.

File: keyboards/whitefox/whitefox.c
Comment (firetech) [reply keyboards/whitefox/whitefox.c]: Apparently, yes. The `palSet...` lines were copied from the old uGFX/visualizer code, I had no idea the cleaner syntax existed. 🙂

Comment (fauxpark) [reply keyboards/whitefox/whitefox.c]: I recall there are some naming conflicts (yet again...) with Kinetis in ChibiOS as some registers have the same names as certain QMK-style pins, so you cannot reference them in `MATRIX_COL/ROW_PINS` in config.h with `#include "config_common.h"` present, otherwise the compiler complains.

**Review by fauxpark — APPROVED** [review]

File: keyboards/whitefox/whitefox.c
Comment (fauxpark) [inline keyboards/whitefox/whitefox.c:91]: Ideally, I would like to see if something like #12947 for the K20x fixes this, before merging.

Comment (firetech) [reply keyboards/whitefox/whitefox.c:91]: As far as my research has lead me to understand, you basically need to wipe the entire MCU and reflash the bootloader to change the "EEPROM" size. 🙁

Comment (fauxpark) [reply keyboards/whitefox/whitefox.c:91]: Then I'm not sure how else this can be solved if we want VIA support (and room for future core settings)... perhaps we should just fall back to the transient EEPROM implementation, that will eliminate the need for this code at least.

Comment (firetech) [reply keyboards/whitefox/whitefox.c:91]: I've done some more investigation and realized that the partitioning code in eeprom_teensy.c isn't actually run on an already partitioned MCU. When I tried increasing `EEPROM_SIZE` locally, I guess my board therefore just crashed due to the firmware configuration not matching the hardware configuration.

The documentation is a bit fuzzy on the subject, but NXP Support claimed in their forum that the relevant flash and registers need to be in "an erased state" before running the PGMPART command. I'm a bit hesitant to try since I don't want to risk bricking my main keyboard, but it might be possible to try skipping the [`FTFL->FCNFG & FTFL_FCNFG_RAMRDY` if statement](https://github.com/qmk/qmk_firmware/blob/master/tmk_core/common/chibios/eeprom_teensy.c#L99) and that way force the PGMPART command to be run. Due to what NXP Support said in the forums, my guess is that it won't work.

Comment (fauxpark) [reply keyboards/whitefox/whitefox.c:91]: > I've done some more investigation and realized that the partitioning code in eeprom_teensy.c isn't actually run on an already partitioned MCU.

As far as I can tell, the original Kiibohd firmware doesn't do any partitioning (so there is no persistent storage, apparently). This code seems to be mainly intended for actual Teensys, which have an external chip doing the bootloading, and according to the PJRC forum thread mentioned in the other PR, it sends the `SETRAM` command to turn the EEPROM functionality off. It also likely does a mass erase first.

And yeah, the manual says:
> After clearing CCIF to launch the Program Partition command, the flash memory module first verifies that the EEPROM Data Size Code and FlexNVM Partition Code in the data flash IFR are erased.

Comment (firetech) [reply keyboards/whitefox/whitefox.c:91]: > As far as I can tell, the original Kiibohd firmware doesn't do any partitioning (so there is no persistent storage, apparently).

Hmm, so it might have been QMK that set my keyboard to 32 bytes of EEPROM, then?

Comment (firetech) [reply keyboards/whitefox/whitefox.c:91]: Sooo, I finally mustered up the courage to try and force the partitioning command to run, and, as expected, it had no effect whatsoever (I did make sure the code actually ran, since `eeprom_initialize` isn't even called if the EEPROM is ready before the first read or write). Reading the manual, it seems like the only way to clear the data flash IFR (after which partitioning should be allowed again) is the "clear all blocks" command, which undoubtedly would remove the bootloader. Not really a safe move...

The transient EEPROM driver might be the way to go for non-Teensy K20x boards... 😢 

I guess crazy users (like me 😛) who _really_ want to use EEPROM could theoretically implement a custom (hackish) keymap level EEPROM driver that reshuffles the eeconfig addresses to more effectively use the 32 available bytes. \*evil laugh*

Comment (fauxpark) [reply keyboards/whitefox/whitefox.c:91]: At least the bootloaders for all Input:Club boards can be found here: https://github.com/kiibohd/controller/releases
`ergodox.v0.5.7.bootloader.bin`

All you need is a J-Link, I suppose.

Comment (firetech) [reply keyboards/whitefox/whitefox.c:91]: How should we go ahead with this? I feel like changing the boards to the transient EEPROM driver might be a bit out of scope for this PR, but without either that or these lines of code, LED Matrix is effectively dead after the second boot (due to `flags` being 0)...

Comment (fauxpark) [reply keyboards/whitefox/whitefox.c:91]: I'm fine to merge this as is, now that I know a bit more about the problem.

Since this is technically a bug in QMK, it should get a proper fix. I would kinda like to fix the code in a separate, breaking changes PR - it would mean users will need to wipe their boards, of course. But, it should only need to be done once.
Failing that, perhaps we can adapt the STM32 emulated eeprom?

Comment (firetech) [reply keyboards/whitefox/whitefox.c:91]: > Since this is technically a bug in QMK, it should get a proper fix. I would kinda like to fix the code in a separate, breaking changes PR - it would mean users will need to wipe their boards, of course. But, it should only need to be done once.

I have some doubts about how many users would be willing (and able) to wipe their boards like that...

If you just increase the default `EEPROM_SIZE` without changing anything else, boards initialized with another value will fail to boot. At least that's how mine behave.

Comment (firetech) [reply keyboards/whitefox/whitefox.c:91]: Hmm. It might even be possible to switch the FlexRAM function over from EEPROM to RAM when an incompatible EESIZE is detected in `SIM->FCFG1`, in effect making the EEPROM driver work using the same code, but behaving transient. I'm gonna do some experimenting. 👨‍🔬

Comment (firetech) [reply keyboards/whitefox/whitefox.c:91]: [Working prototype](https://github.com/firetech/qmk_firmware/commit/kinetis_flexram_fallback)! 😄

I also realized that the "crash" during boot after increasing `EEPROM_SIZE` was just an infinite loop in `flexram_wait()` due to the `EEERDY` flag never being set after an out-of-bounds write. Just making sure the FlexRAM isn't written to outside the partitioned EEPROM size is enough to make the board boot properly. In other words, the entire `if (eeprom_size < EEPROM_SIZE) { ... }` block _can_ be removed (but it might be desired).

If you feel that this would be a valid solution to the problem, I can make a PR of this after #12947 has been merged (would conflict heavily otherwise).

File: keyboards/whitefox/config.h
Comment (fauxpark) [inline keyboards/whitefox/config.h:69]:
```suggestion
#define LED_DISABLE_WHEN_USB_SUSPENDED
```
#13146

File: keyboards/ergodox_infinity/config.h
Comment (fauxpark) [inline keyboards/ergodox_infinity/config.h:70]:
```suggestion
#define LED_DISABLE_WHEN_USB_SUSPENDED
```

**Review by fauxpark — CHANGES_REQUESTED** [review]
One last thing, then I think I can merge this :)

File: keyboards/whitefox/whitefox.c
Comment (fauxpark) [reply keyboards/whitefox/whitefox.c:91]: An issue detailing all of these findings may be a good idea, rather than having it all in this tangentially-related PR discussion.

**Comment by firetech** [comment]
Fixed.

Was considering adding a check of `EEPROM_SIZE` to the eeprom limit hack in matrix_init_kb discussed above, but realized that `EEPROM_SIZE` probably isn't defined in that context...

File: keyboards/whitefox/whitefox.c
Comment (firetech) [reply keyboards/whitefox/whitefox.c:91]: Good point... I'll get one started.

**Review by fauxpark — APPROVED** [review]

**Comment by fauxpark** [comment]
Thanks for all your work!

File: keyboards/whitefox/whitefox.c
Comment (firetech) [reply keyboards/whitefox/whitefox.c:91]: #13164

**Pull request merged**
